### ISSUE TITLE
Adding and removing traits is bugged

### ISSUE THREAD

**Issue by QuantumCoderQC** [armory3d/armory#2289]
**Description**

Removing and adding of a trait works only once. After that, the trait no longer works as expected.

Thanks to @ wuaieyo and @ Henriq on Discord to point out the issue

The test scene contains a default cube, which has 3 traits added to it. 

1. `MyTrait`: Haxe trait that rotates the object on update at a given rate.
2. `NodeTree`: Logic node tree that rotates the object on update at a given rate.
3. `AddRemove`: Logic node tree that adds and removes the above two traits. Keys A, B remove and add "Node Tree". Keys C, D remove and add `MyTrait`.

![image](https://user-images.githubusercontent.com/55564981/128167005-a63a5cd8-175c-4241-9b3c-27f7951c7355.png)
Start with `NodeTree` and `AddRemove` enabled. The cube rotates as expected. Now, on key press A, the trait `NodeTree` is removed and the cube stops rotating. On key press B, the same trait is added back and the cube starts rotating again.

Now, After removing and adding back the trait, the cube no longer rotates. Although from the debug window, it can be seen that the trait is in fact added to the cube. So, removing and adding of a trait works only once regardless of the duration between adding and removing. 

The same is the case with the Haxe trait `MyTrait`. It can also be removed and added back only once. 

Another issue in conjunction with this is if a trait is `exported` and not `enabled`, the trait is still active at start. 
![image](https://user-images.githubusercontent.com/55564981/128166507-3ffc5163-26d6-4479-b6f1-7f1c4feb3bea.png)

This also means that disabling both `export` and `enable`, the trait is never exported and cannot be added at run time.
![image](https://user-images.githubusercontent.com/55564981/128166869-b0221ae8-927b-4cc5-a7e5-032f0e010d50.png)

**To Reproduce**
Use the provided blend file to test.

**Expected behavior**
It must be possible to add and remove traits without limitations, if a trait is exported. If a trait is exported and not enabled, the trait must not be run at start.

**System**
Blender: 2.93 LTS
Armory: Armory v2021.7 ($Id: 8f3d9af20f4056024e616db2c87703a94c568023 $)
OS: Windows 10.

**Test File**
[AddRemoveTraitTest.zip](https://github.com/armory3d/armory/files/6930556/AddRemoveTraitTest.zip)

**Comment by MoritzBrueckner** [comment]
I thought a bit about this issue and I think I have a solution, but before implementing it I want to make sure that it sounds good to you.

The caching of the `Trait` node seems to be correct, as it can be used as a variable node. It is not meant to be a "Get" or "Create" node (however there is the question if it's really needed then). Instead, the `Get Object Trait` node should probably be used for the add/remove trait instead, but currently it only works for removal as the trait has to exist already. So, my idea is to add a checkbox `Create If Missing` to this node that would instantiate the trait if it isn't found. This would solve this issue. If the trait doesn't exist, create it, otherwise use the existing trait.

This might a bit counterintuitive first given the node's name (you wouldn't search for a "get" node if you want to create something), but I think this is a simple and well-working solution. The whole trait system is a bit awkward to work with from the engine perspective, but I think that it has to be this way to be cache-friendly (the trait system uses a [data-oriented design](https://en.wikipedia.org/wiki/Data-oriented_design) approach).

**Comment by QuantumCoderQC** [comment]
Hi, Thanks for looking into the issue.

> Instead, the Get Object Trait node should probably be used for the add/remove trait instead. 

I would disagree with this as the this node can also be used to check if an object even has the mentioned trait and do something if it does/ not. So, I would suggest not to use the "create trait" checkbox on the `Get Object Trait` node.

What would be better, in my opinion, is to use this checkbox in the `Trait` node. 

Or is the checkbox even required in this case? Since when someone uses the `Trait` node, they expect a trait to be returned with that name(if it exists of course) weather or not the trait is cached.

One more, probably cleaner(?) solution I could think of would be to use the `iron.data.Data` to cache the traits. Then we can have `iron.data.Data.getTrait("MyTrait");` and should return the trait regardless of it being cached. 

Please let me know what you think of my thoughts.

**Comment by MoritzBrueckner** [comment]
> I would disagree with this as the this node can also be used to check if an object even has the mentioned trait and do something if it does/ not. So, I would suggest not to use the "create trait" checkbox on the Get Object Trait node.

Wouldn't this be the behaviour if the checkbox is unchecked?

> What would be better, in my opinion, is to use this checkbox in the Trait node.

This could be possible as well, but then the trait node would need to check whether the trait already exists for the calling object, which is a bit against the node's functionality. And what if the currently executing trait wants to access a trait on another object? Then the trait node would need to have another object field, basically transforming it into the `Get Object Trait` node.

> Or is the checkbox even required in this case? Since when someone uses the Trait node, they expect a trait to be returned with that name(if it exists of course) weather or not the trait is cached.

Yes, it is required, because without that this issue happens. The trait node stores a trait that is already instanciated so when re-adding the trait the callbacks are no longer registered. One way around that would be to register the callbacks in `notifyOnAdd`. It's a somewhat weird situation right now because it's not quite clear what the original trait philosophy was, what should happen on instanciation and what on adding the trait?

> One more, probably cleaner(?) solution I could think of would be to use the iron.data.Data to cache the traits. Then we can have iron.data.Data.getTrait("MyTrait"); and should return the trait regardless of it being cached.

Yes, something like this would definitely be great, however a trait can have multiple instances (do they have the same name?) which would make this access more difficult. But it could be possible.

**Comment by QuantumCoderQC** [comment]
> Wouldn't this be the behaviour if the checkbox is unchecked?

Yes, your are correct. But still, I think it will add for some confusion for newer users.

> And what if the currently executing trait wants to access a trait on another object? Then the trait node would need to have another object field, basically transforming it into the Get Object Trait node.

This is a different situation, where you want trait applied on a particular object. But, in general, traits dont necessarily need objects to exist. They need objects only to execute the commands. 

So, in my example, I had a trait which rotates any object it is applied to. Now, I want to be able to start with an empty scene -> spawn a cube -> add this trait to it. 

In this situation, I would naturally use the `Trait` node and not the `Get Object Trait` node since there are no objects when I begin. 

> This could be possible as well, but then the trait node would need to check whether the trait already exists for the calling object, which is a bit against the node's functionality

> Yes, it is required, because without that this issue happens. The trait node stores a trait that is already instanciated so when re-adding the trait the callbacks are no longer registered

I think instantiating a trait is not the job of the `Trait` node. But the `Add Trait To Object` node. One way to shift this responsibility is to change the `Trait` field to `String` field in the node, just like the `Get object Trait` node. That way, we can instantiate the trait and then add it to the object. This way the callbacks should also be registered.

**Comment by MoritzBrueckner** [comment]
> I think instantiating a trait is not the job of the `Trait` node. But the `Add Trait To Object` node. One way to shift this responsibility is to change the `Trait` field to `String` field in the node, just like the `Get object Trait` node. That way, we can instantiate the trait and then add it to the object. This way the callbacks should also be registered.

This is probably the cleanest solution, although it would render the `Trait` node a bit useless because it's simply a `String` node. Should we keep it anyway or deprecate it?

**Comment by QuantumCoderQC** [comment]
We can just keep it. Mabe some use would come up later.
